cql3: expr: make it possible to prepare and evaluate conjunctions

cvybhu opened the pull request:
This PR implements two things:
* Getting the value of a conjunction of elements separated by `AND` using `expr::evaluate`
* Preparing conjunctions using `prepare_expression`

---

`NULL` is treated as an "unkown value" - maybe `true` maybe `false`.
`TRUE AND NULL` evaluates to `NULL` because it might be `true` but also might be `false`.
`FALSE AND NULL` evaluates to `FALSE` because no matter what value `NULL` acts as, the result will still be `FALSE`.
Unset and empty values are not allowed.

Usually in CQL the rule is that when `NULL` occurs in an operation the whole expression becomes `NULL`, but here we decided to deviate from this behavior.
Treating `NULL` as an "unkown value" is the standard SQL way of handing `NULLs` in conjunctions.
It works this way in MySQL and Postgres so we do it this way as well.

The evaluation short-circuits. Once `FALSE` is encountered the function returns `FALSE` immediately without evaluating any further elements.
It works this way in Postgres as well, for example:
`SELECT true AND NULL AND 1/0 = 0` will throw a division by zero error,
 but `SELECT false AND 1/0 = 0` will successfully evaluate to `FALSE`.

avikivity commented:
I would have guessed that short-circuiting is not allowed, but it is.

https://stackoverflow.com/questions/789231/is-the-sql-where-clause-short-circuit-evaluated

avikivity commented on cql3/expr/prepare_expr.cc:
```diff
@@ -875,6 +875,53 @@ test_assignment_function_call(const cql3::expr::function_call& fc, data_dictiona
+    bool all_terminal = true;
+    for (const expression& child : conj.children) {
+        std::optional<expression> prepared_child =
+            try_prepare_expression(child, db, keyspace, schema_opt, child_receiver);
```
You could adjust the receiver to say "3rd element of conjunction".

But I'm against it, it's not useful. One day we'll track line+column information and then the receiver will contain that information.

cvybhu replied:
I'm leaving it as is then

avikivity commented on cql3/expr/prepare_expr.cc:
```diff
@@ -875,6 +875,53 @@ test_assignment_function_call(const cql3::expr::function_call& fc, data_dictiona
+        std::optional<expression> prepared_child =
+            try_prepare_expression(child, db, keyspace, schema_opt, child_receiver);
+        if (!prepared_child.has_value()) {
+            return std::nullopt;
```
I think you can just fail here. Returning nullopt makes sense if we can try again and infer the type (for example, `? = X` will fail when we try `?` first but succeed after we get the type of `X`). But here, the input to a conjunction has to be bool.

(btw I think it's impossible for try_prepare_expr to return nullopt if the receiver is not nullptr).

cvybhu replied:
done

avikivity commented:
Looks good, find a minor thing (that I think can't happen in practice, but better to tighten).

cvybhu commented:
v2:
* Throw an exception when unable to infer the type of a conjunction's child. The message is based on the one `prepare_expression` produces when `try_prepare_expression` is unable to infer the type.
* Fixed typo, added missing `in` inside `It works this way in MySQL`

avikivity commented on cql3/expr/prepare_expr.cc:
```diff
@@ -875,6 +875,53 @@ test_assignment_function_call(const cql3::expr::function_call& fc, data_dictiona
+        std::optional<expression> prepared_child =
+            try_prepare_expression(child, db, keyspace, schema_opt, child_receiver);
+        if (!prepared_child.has_value()) {
+            throw exceptions::invalid_request_exception(fmt::format("Could not infer type of {}", child));
```
The exception is wrong, since receiver != nullptr we're not trying to infer the type. But it can't happen anyway.

cvybhu replied:
Well if the result of `try_prepare_expression` is `nullopt` that means that type inference failed.
It should never fail, but if it does then I think it's the right message.
I could change it to `on_internal_error`, but it's probably not worth it to spend time one it.

nyh commented:
@cvybhu if I understand correctly, this adds to Scylla new types of expressions that weren't previously supported by Cassandra. If this is the case, I think we should start to document these differences, perhaps in docs/cql/cql-extensions.md.

Personally I would also have liked to see cql tests for such CQL syntax features (test/cql-pytest or test/cql or dtest or ...) and not just C++ unit tests because we can't run C++ unit tests against Cassandra, and we are running a risk that one of the features we thought we added, Cassandra actually supports - but differently. All the NULL stuff is an obvious risk of deviating from Cassandra, which we do, but we should know that we do, and document these differences in tests and/or the cql-extensions document.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/3482/

avikivity commented:
> @cvybhu if I understand correctly, this adds to Scylla new types of expressions that weren't previously supported by Cassandra. If this is the case, I think we should start to document these differences, perhaps in docs/cql/cql-extensions.md.
> 
> Personally I would also have liked to see cql tests for such CQL syntax features (test/cql-pytest or test/cql or dtest or ...) and not just C++ unit tests because we can't run C++ unit tests against Cassandra, and we are running a risk that one of the features we thought we added, Cassandra actually supports - but differently. All the NULL stuff is an obvious risk of deviating from Cassandra, which we do, but we should know that we do, and document these differences in tests and/or the cql-extensions document.

AND expressions are supported by Cassandra.

nyh commented:
> AND expressions are supported by Cassandra.

So I'm still at a loss what this patch does... Does it re-implement something we already supported, just in a different way? Does the re-implementation add support perhaps for more complicated expressions that weren't allowed before or it's exactly the same expressions which are supported?

These are probably stupid questions, but I feel like I'm really missing the context of this pull request and how it fits into the whole CQL rewrite story.

Also, if this a re-implementation of something which we already supported, how do we test that we didn't suffer a regression - i.e., that some AND expression that worked previously now works incorrectly or differently? The new C++ tests only test the new expression framework, so if I understand correctly cannot be used to compare behavior to the previous implementation :-( Do we have any dtest tests for this feature? (I'm not a big fan of dtest, I prefer cql-pytest, but if we have dtests for this, it's good enough).

avikivity commented:
> > AND expressions are supported by Cassandra.
> 
> So I'm still at a loss what this patch does...

I guess the cover letter forgot to mention why it does the things it does.

> Does it re-implement something we already supported, just in a different way? Does the re-implementation add support perhaps for more complicated expressions that weren't allowed before or it's exactly the same expressions which are supported?

There is a long arc here. CQL expressions used to be represented by multiple object-oriented trees (one for each role - WHERE clause atom, selector, LWT IF clause atoms). It is now represented by a passive structure (expression) supported by a unified set of functions (prepare_expr, evaluate).

This series make one more step from the old-world to the new world by making evaluate() and prepare_expr() able to understand conjunctions, but it doesn't yet change the callers to use them.

> 
> These are probably stupid questions, but I feel like I'm really missing the context of this pull request and how it fits into the whole CQL rewrite story.
> 
> Also, if this a re-implementation of something which we already supported, how do we test that we didn't suffer a regression - i.e., that some AND expression that worked previously now works incorrectly or differently?

Since this just adds new code and doesn't replace the old code, nothing in expression evaluation has yet to change.

Once we do switch over the implementation, we still expect nothing to change. At that time it will be important to worry about the regression suite. I have confidence in the expression tests and in the coverage provided by our existing suite, but that's a matter of opinion.

>  The new C++ tests only test the new expression framework, so if I understand correctly cannot be used to compare behavior to the previous implementation :-( Do we have any dtest tests for this feature? (I'm not a big fan of dtest, I prefer cql-pytest, but if we have dtests for this, it's good enough).

There's really no feature here. We're changing the mechanics of how AND is prepared and evaluated, not what AND does.

AND is simple enough that I trust we won't have screwed it up.

Once the this is done, we can truly make changes and relax the grammar. Like allowing `SELECT ... WHERE a_boolean_column` (rather than requiring a relational operator) or `SELECT a = b AND b = c` (rather than not allowing it). Any place CQL accepts "some form of an expression" will be changed to accept "an expression".
